### ISSUE TITLE
Failure to decompile

### ISSUE THREAD

**Issue by guygodin** [icsharpcode/ILSpy#1099]
With the recent versions of ILSpy, I get the following crash when selecting some class members:
```
System.NotSupportedException: Specified method is not supported.
   at Mono.Cecil.Cil.InstructionOffset.get_Offset()
   at Mono.Cecil.Cil.CodeReader.ReadScope(ScopeDebugInformation scope)
   at Mono.Cecil.Cil.CodeReader.ReadDebugInfo()
   at Mono.Cecil.Cil.CodeReader.ReadMethodBody()
   at Mono.Cecil.Cil.CodeReader.ReadMethodBody(MethodDefinition method)
   at Mono.Cecil.MethodDefinition.<>c.<get_Body>b__41_0(MethodDefinition method, MetadataReader reader)
   at Mono.Cecil.ModuleDefinition.Read[TItem,TRet](TRet& variable, TItem item, Func`3 read)
   at Mono.Cecil.MethodDefinition.get_Body()
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.CollectNamespacesForDecompilation(IEnumerable`1 memberDefinitions, HashSet`1 namespaces, HashSet`1 visited)
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.Decompile(IList`1 definitions)
   at ICSharpCode.ILSpy.CSharpLanguage.DecompileMethod(MethodDefinition method, ITextOutput output, DecompilationOptions options)
   at ICSharpCode.ILSpy.TextView.DecompilerTextView.DecompileNodes(DecompilationContext context, ITextOutput textOutput)
   at ICSharpCode.ILSpy.TextView.DecompilerTextView.<>c__DisplayClass33_0.<DecompileAsync>b__0()
```

If I restart ILSpy and try the same member again, it works but once the error occurs, it won't be able to decompile anything else and a restart is required.

**Comment by guygodin** [comment]
It seems like unchecking "Use variable names from debug symbols" will fix the issue. But sine it is checked by default, a lot of folks will run into that.

**Comment by siegfriedpammer** [comment]
Can you provide an assembly that causes this problem? I have been unable to reproduce this. Thank you!

**Comment by guygodin** [comment]
Someone posted repro steps here: https://github.com/icsharpcode/ILSpy/issues/1054
Let me know if this doesn't work and I can upload a dll with repro steps

**Comment by siegfriedpammer** [comment]
The problem mentioned in #1054 is a different one, but I will try to reproduce this problem with the other DLL.

**Comment by guygodin** [comment]
[SiliconStudio.Xenko.zip](https://github.com/icsharpcode/ILSpy/files/1830787/SiliconStudio.Xenko.zip)

Go to SiliconStudio.Xenko.FourCC, right-click Analyze, Used by, then go to the GetHashCode method, then expand the Used By method.

**Comment by siegfriedpammer** [comment]
The above exception has nothing to do with analyzers. It should occur while trying to decompile something. Analyzers do not decompile anything. I cannot reproduce the analyzer exception either. (See my comment in #1054)

**Comment by siegfriedpammer** [comment]
Copied from the other issue:

> Keep in mind that the analyzers work on the full assembly list, so any loaded assembly could be the culprit. Please try to find the culprit by removing other assemblies one after the other. (But make a copy of %APPDATA%\ICSharpCode\ILSpy.xml beforehand, so you can always go back to the erroneous state!)

**Comment by guygodin** [comment]
I can repro the crash by deleting the %APPDATA%\Roaming\ICSharpCode\ILSpy.xml, then installing a fresh copy of ILSpy and adding the assembly I just linked above and following the repro steps.

**Comment by guygodin** [comment]
[SiliconStudio.Xenko.zip](https://github.com/icsharpcode/ILSpy/files/1832450/SiliconStudio.Xenko.zip)

Here you go

**Comment by siegfriedpammer** [comment]
Thanks, but still no luck :( ... Can you still reproduce the problem, if you add the dll + pdb to a separate folder and open it from there? I suspect some reference gets automatically loaded on your side.

**Comment by guygodin** [comment]
Yeah it occurs even when copying the dll and pdb into a seperate folder. When adding the dll to ILSpy, it adds references (in light blue) underneath:

mscorlib (2.0.0.0): C:\Windows\Microsoft.NET\Framework64\v2.0.50727\mscorlib.dll
System (4.0.0.0): C:\Windows\Microsoft.NET\Framework64\v4.0.30319\System.dll
System.Core (4.0.0.0): C:\Windows\Microsoft.NET\Framework64\v4.0.30319\System.Core.dll

I have ILSpy setup to use C# 7.2 / VS 2017

**Comment by siegfriedpammer** [comment]
OK... I am out of ideas...

**Comment by siegfriedpammer** [comment]
Can you try getting the source and debugging it yourself? Otherwise we could try Skype or Teamviewer... any other options?

**Comment by n9** [comment]
I have hit a similar problem, when I click "Used By" on MvcJsonOptions.
Assembly: `Microsoft.AspNetCore.Mvc.Formatters.Json, Version=2.0.3.0, Culture=neutral, PublicKeyToken=adb9793829ddae60`
Class: `Microsoft.AspNetCore.Mvc.MvcJsonOptions`
ILSpy version: `3.0.1.3459`.

**Comment by n9** [comment]
I cannot copy the full exception from the analyzer (no context menu, Ctrl+C not working). There is a screenshot:

![image](https://user-images.githubusercontent.com/1651867/37825260-7679322c-2e8f-11e8-8b52-d44934fac02b.png)

**Comment by n9** [comment]
I have tried to build the current master and the output is following:
![image](https://user-images.githubusercontent.com/1651867/37825635-d21bd3fe-2e90-11e8-9187-459e3105b9b4.png)

(Used By on master works.)

**Comment by siegfriedpammer** [comment]
I suspect this is a threading issue. The only way to properly fix this seems to switch Cecil for SRM, which will be done in ILSpy 4.0.

**Comment by siegfriedpammer** [comment]
Can you try the latest nightly build? (Get it from here: https://ci.appveyor.com/project/icsharpcode/ilspy) Thank you!

**Comment by n9** [comment]
Still building...

**Comment by twenzel** [comment]
@siegfriedpammer There are less occurences, but problem still exists.

**Comment by siegfriedpammer** [comment]
Can you post a stack trace? Thank you!

**Comment by twenzel** [comment]
using version `3.1.0.3646-rc`

```
 ---> System.NotSupportedException: Specified method is not supported.
   at Mono.Cecil.Cil.InstructionOffset.get_Offset() in offset 56
   at Mono.Cecil.Cil.CodeReader.ReadScope(ScopeDebugInformation scope) in offset 31
   at Mono.Cecil.Cil.CodeReader.ReadDebugInfo() in offset 64
   at Mono.Cecil.Cil.CodeReader.ReadMethodBody() in offset 151
   at Mono.Cecil.Cil.CodeReader.ReadMethodBody(MethodDefinition method) in offset 26
   at Mono.Cecil.MethodDefinition.<>c.<get_Body>b__41_0(MethodDefinition method, MetadataReader reader) in offset 0
   at Mono.Cecil.ModuleDefinition.Read[TItem,TRet](TRet& variable, TItem item, Func`3 read) in offset 64
   at Mono.Cecil.MethodDefinition.get_Body() in offset 102
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DecompileBody(MethodDefinition methodDefinition, IMethod method, EntityDeclaration entityDecl, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 38
-- continuing with outer exception (ICSharpCode.Decompiler.DecompilerException) --
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DecompileBody(MethodDefinition methodDefinition, IMethod method, EntityDeclaration entityDecl, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 805
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DoDecompile(MethodDefinition methodDefinition, IMethod method, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 78
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.Decompile(IList`1 definitions) in offset 331
   at ICSharpCode.ILSpy.CSharpLanguage.DecompileMethod(MethodDefinition method, ITextOutput output, DecompilationOptions options) in offset 115
   at ICSharpCode.ILSpy.TextView.DecompilerTextView.DecompileNodes(DecompilationContext context, ITextOutput textOutput) in offset 61
   at ICSharpCode.ILSpy.TextView.DecompilerTextView.<>c__DisplayClass33_0.<DecompileAsync>b__0() in offset 36
```

It occurs only after "analyse" a method, expand the "Used by" node and selecting any other method to decompile.

**Comment by siegfriedpammer** [comment]
Yes, the exception in the analyzers cannot be fixed easily... (without rewriting them) removing/renaming the broken PDB is the only workaround.

There are two separate issues:
1) exception during decompilation (see first post of this issue) FIXED
2) exception in the analyzer (see other posts) WONT FIX (for now)

**Comment by siegfriedpammer** [comment]
With the move to SRM and the reimplementation of the analyzers, problems like this should disappear. If you still encounter any problems, please open a new issue.
